Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -551,38 +580,330 @@ ErrorOr<void> SDHostController::transaction_control_with_data_transfer_using_the
+            if (!retry_with_timeout([this]() { return m_registers->interrupt_status.command_complete; })) {
+                // FIXME: Send an Abort signal
+                return EIO;
+            }
+            // (9) Write 1 to the Command Complete in the Normal Interrupt Status register to clear this bit.
+            m_registers->interrupt_status.raw = command_complete;
+            // FIXME: (10) Read Response register and get necessary information of the issued command
+            // (11) Wait for the Transfer Complete Interrupt and ADMA Error Interrupt.
+            if (!retry_with_timeout([this]() { return m_registers->interrupt_status.command_complete || m_registers->interrupt_status.adma_error; })) {
+                // FIXME: Send an Abort signal
+                return EIO;
+            }
+            // (12) If Transfer Complete is set to 1, go to Step (13)
+            if (m_registers->interrupt_status.command_complete) {
+                // (13) Write 1 to the Transfer Complete Status in the Normal Interrupt Status register to clear this bit.
+                m_registers->interrupt_status.raw = command_complete;
+
+            }
+            //      else if ADMA Error Interrupt is set to 1, go to Step (14).
+            else if (m_registers->interrupt_status.adma_error) {
+                // (14) Write 1 to the ADMA Error Interrupt Status in the Error Interrupt Status register to clear this bit.
+                m_registers->interrupt_status.adma_error = 1;
+                // (15) Abort ADMA operation. SD card operation should be stopped by issuing abort command. If
+                //      necessary, the Host Driver checks ADMA Error Status register to detect why ADMA error is
+                //      generated
+                TODO();
+            } else {
+                VERIFY_NOT_REACHED();
+            }
+
+            blocks_read_total += blocks_read;
+            dbgln("Read {} blocks ({} bytes) to {:p}", blocks_read, blocks_read * block_len, m_dma_region->vaddr().get() + PAGE_SIZE);
+            dbgln("Data:\n{}", ReadonlyBytes(bit_cast<void const*>(m_dma_region->vaddr().get() + PAGE_SIZE), blocks_read * block_len));
+            dbgln("{} blocks remaining in read-op", block_count - blocks_read_total);
+            TRY(out.write(bit_cast<void const*>(m_dma_region->vaddr().get() + PAGE_SIZE), write_offset, blocks_read * block_len));
+            write_offset = read_offset;
+        }
+        dbgln("Stopped reading");
+        break;
+    }
+    default:
+        VERIFY_NOT_REACHED();
+    }
     return {};
 }
 
 ErrorOr<void> SDHostController::read_block(Badge<SDMemoryCard>, u32 block_address, u32 block_count, UserOrKernelBuffer out)
 {
     VERIFY(is_card_inserted());
 
-    if (block_count > 1) {
+    using enum OperatingMode;
+    switch (m_mode) {
+    case OperatingMode::ADMA2_32:
+    case OperatingMode::ADMA2_64:
+        return read_block_adma2(block_address, block_count, out);
```

Comment:
it feels you could naturally split the big method here per mode… 🤔

Reply:
Refactored it a bit and should be nicer to read now

Reply:
it is! 😃

yet i wonʼt try to understand the logic nor cross-check the spec; excuses! 😅